crates/network-scanner/src/ping.rs

```diff
@@ -1,34 +1,66 @@
-use std::net::Ipv4Addr;
+use std::{
+    mem::MaybeUninit,
+    net::{Ipv4Addr, SocketAddr},
+};
 
-use network_scanner_net::tokio_raw_socket::TokioRawSocketStream;
+use anyhow::Context;
+use network_scanner_net::tokio_raw_socket::TokioRawSocket;
 use network_scanner_proto::icmp_v4;
 
-use anyhow::Context;
-use tokio::io::{AsyncReadExt, AsyncWriteExt};
 use tracing::trace;
 
+#[macro_export]
+macro_rules! create_echo_request {
+    () => {{
+        let time = std::time::SystemTime::now()
+            .duration_since(std::time::UNIX_EPOCH)
+            .map_err(|e| anyhow::anyhow!(e))?
+            .as_secs();
+
+        let echo_message = network_scanner_proto::icmp_v4::Icmpv4Message::Echo {
+            identifier: 0,
+            sequence: 0,
+            payload: time.to_be_bytes().to_vec(),
+        };
+
+        let packet = network_scanner_proto::icmp_v4::Icmpv4Packet::from_message(echo_message);
+        (packet, time.to_be_bytes().to_vec())
+    }};
+}
+
 pub async fn ping(ip: Ipv4Addr) -> anyhow::Result<()> {
-    let mut socket = TokioRawSocketStream::connect(ip)
+    let socket = TokioRawSocket::new(
+        socket2::Domain::IPV4,
+        socket2::Type::RAW,
+        Some(socket2::Protocol::ICMPV4),
+    )
+    .with_context(|| format!("failed to create tokio raw socket"))?;
+
+    let addr = SocketAddr::new(ip.into(), 0);
+
+    let (packet, verifier) = create_echo_request!();
+
+    socket
+        .send_to(&packet.to_bytes(true), socket2::SockAddr::from(addr))
         .await
-        .with_context(|| format!("could not connect to {ip}"))?;
+        .with_context(|| format!("Failed to send packet to {}", ip))?;
 
-    let time = std::time::SystemTime::now()
-        .duration_since(std::time::UNIX_EPOCH)
-        .context("cannot access system time")?
-        .as_secs();
+    let mut buffer = [MaybeUninit::uninit(); icmp_v4::ICMPV4_MTU];
 
-    let echo_message = network_scanner_proto::icmp_v4::Icmpv4Message::Echo {
-        identifier: 0,
-        sequence: 0,
-        payload: time.to_be_bytes().to_vec(),
-    };
+    let (size, _) = socket
+        .recv_from(&mut buffer)
+        .await
+        .with_context(|| format!("Failed to receive packet from {}", ip))?;
+
+    let inited_buf = buffer[..size].as_ref();
 
-    let packet = icmp_v4::Icmpv4Packet::from_message(echo_message);
-    socket.write(&packet.to_bytes(true)).await?;
-    let mut buffer = [0u8; icmp_v4::ICMPV4_MTU];
-    let size = socket.read(&mut buffer).await.map_err(|e| anyhow::anyhow!(e))?;
+    let buffer = inited_buf
+        .iter()
+        .map(|u| unsafe { u.assume_init() })
```

Comment:
btw, I could not find anything that for convert MaybeUinit<u8> to regular u8 without using unsafe keyword, if you are aware of better way of doing this, let me know.

Reply:
There is no safe way to do this without using unsafe code. MaybeUninit is a building block specifically for writing unsafe code.

However, here is a better way to write this code:
```diff
diff --git a/crates/network-scanner/src/ping.rs b/crates/network-scanner/src/ping.rs
index 3044309..6cbbf0e 100644
--- a/crates/network-scanner/src/ping.rs
+++ b/crates/network-scanner/src/ping.rs
@@ -91,14 +91,11 @@ pub fn block_ping(ip: Ipv4Addr) -> anyhow::Result<()> {
         .recv_from(&mut buffer)
         .with_context(|| format!("failed to receive packet from {}", ip))?;

-    let inited_buf = buffer[..size].as_ref();
-
-    let buffer = inited_buf
-        .iter()
-        .map(|u| unsafe { u.assume_init() })
-        .collect::<Vec<u8>>();
+    // SAFETY: `recv_from` returns the number of bytes written into the buffer, so the `size` first
+    // elements are in an initialized state.
+    let filled = unsafe { assume_init(&buffer[..size]) };

-    let packet = icmp_v4::Icmpv4Packet::parse(&buffer[..size]).with_context(|| format!("cannot parse icmp packet"))?;
+    let packet = icmp_v4::Icmpv4Packet::parse(filled).with_context(|| format!("cannot parse icmp packet"))?;

     if let icmp_v4::Icmpv4Message::EchoReply { payload, .. } = packet.message {
         if payload != verifier {
@@ -110,3 +107,15 @@ pub fn block_ping(ip: Ipv4Addr) -> anyhow::Result<()> {
         anyhow::bail!("received non-echo reply");
     }
 }
+
+/// Assume the `buf`fer to be initialised.
+///
+/// # Safety
+///
+/// It is up to the caller to guarantee that the MaybeUninit<T> elements really are in an initialized state.
+/// Calling this when the content is not yet fully initialized causes undefined behavior.
+// TODO: replace with `MaybeUninit::slice_assume_init_ref` once stable.
+// https://github.com/rust-lang/rust/issues/63569
+unsafe fn assume_init(buf: &[MaybeUninit<u8>]) -> &[u8] {
+    &*(buf as *const [MaybeUninit<u8>] as *const [u8])
+}
```

Assumptions made by the unsafe code should be clearly documented. I also vendored a function from the standard library which is not yet stabilized in latest Rust version.

Reply:
thank you, this looks much better!

Reply:
updated